### ISSUE TITLE
Definite documentation on how to write custom binding and triggers, and troubleshoot section

### ISSUE THREAD

**Issue by Kryptos-FR** [Azure/azure-functions-dotnet-worker#2092]
Hello, I have a hard time making anything work in the isolated worker model. All the available doc is about the in-process model, which is ironic since Microsoft is pushing us towards the other one.

Where can I find a definitive documentation on how to write custom bindings and triggers for the isolated worker?

For example, I just want to be able to inject a class into my functions. But the only place it works is on the class constructor, while I want one per-function call.

Pseudo code:
```csharp
// Types to inject
public interface IMyType {}
public class MyType : IMyType {}

// Register them
[assembly: WorkerExtensionStartup(typeof(MyExtensionStartup))]
public sealed class MyExtensionStartup : WorkerExtensionStartup
{
    public override void Configure(IFunctionsWorkerApplicationBuilder applicationBuilder)
    {
        applicationBuilder.Services.AddSingleton(typeof(IMyType), typeof(MyType));
    }
}

// Function
public class MyFunction
{
    private readonly IMyType _my;
    public MyFunction(IMyType> my)
    {
        _my = my; // OK
    }
    
    [Function("Try_Get")]
    public async Task<HttpResponseData> Get(
        [HttpTrigger(AuthorizationLevel.Function, "get", Route = null)] HttpRequestData request,
        CancellationToken token,
        IMyType? my = null
    )
    {
        // my is always null
        return request.CreateResponse(HttpStatusCode.OK);
    }
}
```

I also tried using an attribute but I don't know how to make it work and there is no documentation for that:

```csharp
[AttributeUsage(AttributeTargets.Parameter)]
public class MyBindingAttribute : BindingAttribute
{

}

// In MyFunction class

    [Function("Try_Get")]
    public async Task<HttpResponseData> Get(
        [HttpTrigger(AuthorizationLevel.Function, "get", Route = null)] HttpRequestData request,
        CancellationToken token,
        [MyBinding] IMyType? my = null
    )
```

Using an "input" binding instead brings an error:

```csharp
[AttributeUsage(AttributeTargets.Parameter)]
public class MyBindingInputAttribute : InputBindingAttribute
{

}

// In MyFunction class

    [Function("Try_Get")]
    public async Task<HttpResponseData> Get(
        [HttpTrigger(AuthorizationLevel.Function, "get", Route = null)] HttpRequestData request,
        CancellationToken token,
        [MyBindingInput] IMyType? my = null
    )
```

> The 'Try_Get' function is in error: The binding type(s) 'myBinding' are not registered. Please ensure the type is correct and the binding extension is installed.

Would be nice to have a documentation page to troubleshoot those errors. It is pointless to report them if there is now way to find a solution.

**Comment by gallivantor** [comment]
Agreed. There seems to be no documentation at all for how to build a custom binding with the isolated worker model. The best I could find was that building on top of your example you need to build a class that implements `IInputConverter` and then attach it to the InputAttribute using this:
`[InputConverter(typeof(MyInputConverter))]`

But that is also not enough to get the binding registered, I get the same error message as you:

> The 'Try_Get' function is in error: The binding type(s) 'myBinding' are not registered. Please ensure the type is correct and the binding extension is installed.

**Comment by fabiocav** [comment]
@Kryptos-FR can you share more details about the scenario you expect to handle with your custom binding? I'm assuming the one you've shared above is a simple example, so it would be good to understand your concrete requirement.

The reason why I'm asking is because custom bindings, and particularly triggers, have very limited support in Functions (this is not specific to isolated, as it applies to the in-proc model as well), and there are new extensibility options exposed by isolated that we might be able to recommend if we better understand your scenario.

**Comment by Kryptos-FR** [comment]
@fabiocav  Let me setup a small repo to illustrate our use. I'll edit this comment once it's ready.

**Comment by mikoskinen** [comment]
Ended up in this issue after trying to find documentation on how to convert our existing WebJobs based function trigger to work with this new model.

Let's say that I would like to develop a custom trigger that works with something like Apache Pulsar or NATS or something which isn't available out of the box. Where do I start? For example I have a custom NATS trigger working but it uses the webjob based sdk and now with the new model, there's this:

_Error AZFW0001 : The attribute 'NatsTriggerAttribute' is a WebJobs attribute and not supported in the .NET Worker (Isolated Process). (https://aka.ms/azfw-rules?ruleid=AZFW0001)_

Is there a migration guide about this subject?

Thanks in advance.

**Comment by Kryptos-FR** [comment]
@fabiocav do you have any guidance on the subject, especially regarding the sample I provided? What are the others ways you mentioned?

**Comment by mattramsay-ice** [comment]
Currently encountering the same black hole regarding documentation on custom trigger bindings in the world of isolated worker functions.

It's currently blocking us from migrating to the isolated worker model.

Other than a couple of articles and trying to reverse engineer the existing worker extensions for OOTB triggers/input/output bindings.

**Comment by ivelten** [comment]
I am having the same problem here.

Our company uses in-proc input and output bindings that we want to migrate to the isolated worker, but the documentation we found so far is not enough, especially to help understand how to create and use custom output bindings.

**Comment by jaliyaudagedara** [comment]
Stumbled upon this.

Any update guys?

**Comment by perum** [comment]
This is a serious problem... We use custom INPUT bindings to parse auth tokens/Claims.
To convert to net8 / isolated, these need to work, and there is NO documentation, or even mention of a breaking change.

**Comment by Joseph-Steven-S** [comment]
Hello,

Is there any update on this?

**Comment by kkdubey** [comment]
Hello,

Any update on this issue?
Is there any document to implement Custom attribute trigger?

**Comment by OzBob** [comment]
Related question on [Stack Overflow on how to migrate/use Custom Binding and Authorization Attribute](https://stackoverflow.com/questions/77223236/custom-authorization-in-azure-functions-v4-with-net-7-isolated#:~:text=Instead%20of%20using%20FunctionExecutingContext,%20you%20can).

An attempt to Blog about it by @maartenba [2021 Custom bindings with Azure Functions .NET Isolated Worker](https://blog.maartenballiauw.be/post/2021/06/01/custom-bindings-with-azure-functions-dotnet-isolated-worker.html) was only successful for **string** inputs, after some hacks on the csproj file and extensions.json, with dependencies on isolated worker runtime having access to NuGet to download packages.

Isolated Worker docs, blogged [on techcommunity.microsoft.com by @cloudmelon](https://techcommunity.microsoft.com/t5/apps-on-azure-blog/announcing-the-net-7-for-azure-functions-isolated-worker-process/ba-p/3671398) are vaguley helpful till they are not [here](https://learn.microsoft.com/en-us/azure/azure-functions/dotnet-isolated-process-guide?tabs=windows#extension-packages)

**Comment by sean09lee** [comment]
Any update on this? I am also finding that having attributes in conjuction with middleware would make my code much cleaner.

For example...

`TestFunctions.cs`
```
[AuthorizationAtribute([Permissions.GlobalAdmin])]
[Function("test1")]
public Task<IActionResult> Run([HttpTrigger(AuthorizationLevel.Anonymous, "get", "post", Route = "test1")] HttpRequestData req){}

[AuthorizationAtribute([Permissions.IndividualUser])]
[Function("test2")]
public Task<IActionResult> Run([HttpTrigger(AuthorizationLevel.Anonymous, "get", "post", Route = "test2")] HttpRequestData req){}
```

`AuthorizationAttribute.cs`
```
[AttributeUsage(AttributeTargets.All)]
public class AuthorizationAttribute(string[] requiredPermissions) : Attribute
{
  // Check user permissions against function required permissions
  public Task<bool> CheckPermissions(){}
}
```

**Comment by anandbu** [comment]
Anyone has found solution to this problem yet? Not able to create perfect InputBinding Trigger, lot of classes are internal to the core, InputConvertor gives error as Binding(s) not registered. 
It is a lot of effort to change the code to remove IAsyncCollector, it should have come part of the new Extensions actually.

**Comment by RohitRanjanMS** [comment]
Hi @Kryptos-FR , we have it documented here.

**Comment by OzBob** [comment]
@RohitRanjanMS please add a url to the docs in your comment.

Currently the [mslesrn Azure Functions triggers and bindings concepts](https://learn.microsoft.com/en-us/azure/azure-functions/functions-triggers-bindings?tabs=isolated-process%2Cnode-v4%2Cpython-v2&pivots=programming-language-csharp#custom-bindings) doc points to a 2020 sample code on
[Github](https://github.com/Azure/azure-webjobs-sdk/wiki/Creating-custom-input-and-output-bindings). That doc doesn't even mention Queue Input binding only output.

**Comment by RohitRanjanMS** [comment]
yeah, I was referring to that doc, let me follow-up with the team.

**Comment by OzBob** [comment]
@RohitRanjanMS @Kryptos-FR @ananduchagawkar @sean09lee  please add your voice to the community post requesting more docs on custom attributes:   https://github.com/Azure/azure-functions-dotnet-worker/discussions/2734 

Also, as per Maarten Balliauw "custom bindings need to run in the host" and "the host itself is no longer your concern" and finally "custom bindings that target the .NET Isolated Worker [need to be from] a NuGet package".

To make a custom binding, first make it in a separate project, nuget it, publish it, reference it in the AzFn project

**Comment by Kryptos-FR** [comment]
The issue I have is not related to having a nuget package or not. See my example project. I was already doing that.

**Comment by bretthacker** [comment]
Adding my .02: this is maddening and sadly becoming par for the course with Microsoft SDKs. I used INameResolver from WebJobs in my in-proc function to enable this for dynamic queue names. These are mandatory to assign the correct queues when deploying in the SDLC ("dev-queue1', "int-queue1", etc.). There is nothing comparable in isolated mode.

I seriously question whether any of these library devs deal with real-world scenarios or if they're just regurgitating what they learned in some sophomore CS coding class, from a professor also lacking real-world experience.

**Comment by OzBob** [comment]
From: [functions-bindings-storage-queue](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-storage-queue?tabs=isolated-process%2Cextensionv5%2Cextensionv3&pivots=programming-language-csharp#host-json):

> This version allows you to bind to types from [Azure.Storage.Queues](https://learn.microsoft.com/en-us/dotnet/api/azure.storage.queues).

> This version supports configuration of triggers and bindings through [.NET Aspire integration](https://learn.microsoft.com/en-us/azure/azure-functions/dotnet-aspire-integration#connection-configuration-with-aspire).

> Add the extension to your project by installing the [NuGet package](https://www.nuget.org/packages/Microsoft.Azure.Functions.Worker.Extensions.Storage.Queues), version 5.x.


again: **"bind to types from Azure.Storage.Queues"**

@Kryptos-FR  I am unsure if this goes any way towards address migration of your [custom input binding in your sample project repo](https://github.com/Kryptos-FR/azure-functions-dotnet-worker-2092/blob/main/src/Functions.Isolated/Samples/SFTPSample.cs#L22C10-L22C21) to the isolated model?

Does the [ActiveMQ custom trigger (in-process/isolated source code repo)](https://github.com/arkiaconsulting/akc-azure-functions-worker-extensions-activemq) provide any code the could assist?

**Comment by drdamour** [comment]
> Adding my .02: this is maddening and sadly becoming par for the course with Microsoft SDKs. I used INameResolver from WebJobs in my in-proc function to enable this for dynamic queue names. These are mandatory to assign the correct queues when deploying in the SDLC ("dev-queue1', "int-queue1", etc.). There is nothing comparable in isolated mode.
> 
> I seriously question whether any of these library devs deal with real-world scenarios or if they're just regurgitating what they learned in some sophomore CS coding class, from a professor also lacking real-world experience.

Not trying to sidestep the issue of lack of dynamic  binding resolution…which if not present in isolated is a huge problem fornus too.

but for different queue names across sdlc cant you use envar references in the attribute vals %name-of-en-var% in your case?
